proposals/0075-precompile-for-secp256r1-sigverify.md

```diff
@@ -0,0 +1,382 @@
+   - The special value `0xFFFF` means "current instruction"
+   - If the index is invalid, return Error
+2. Return `length` bytes starting from `offset`
+   - If this exceeds the `instruction_data` length, return Error
+
+Note that fields (offsets) can overlap, for example the same public key or
+message can be referred to by multiple instances of `Secp256r1SignatureOffsets`.
+
+If the precompile `verify` function returns any error, the whole transaction
+should fail. Therefore, the type of error is irrelevant and is left as an
+implementation detail.
+
+The instruction processing logic must follow the pseudocode below:
+
+```
+/// `data` is the secp256r1 program's instruction data. `instruction_datas` is
+/// the full slice of instruction datas for all instructions in the transaction,
+/// including the secp256r1 program's instruction data.
+
+/// length_of_data is the length of `data`
+
+/// SERIALIZED_OFFSET_STRUCT_SIZE is the length of the serialized
+/// Secp256r1SignatureOffsets struct
+
+/// SERIALIZED_PUBLIC_KEY_LENGTH and SERIALIZED_SIGNATURE_LENGTH represent the 
+/// length of the serialized public key and signature respectively
+
+function verify() {
+  if length_of_data == 0 {
+    return Error
+  }
+  num_signatures = data[0]
+  if num_signatures > 8 {
+    return Error
+  }
+  if num_signatures == 0 && length_of_data > 1 {
+    return Error
+  }
+  if length_of_data < (num_signatures * SERIALIZED_OFFSET_STRUCT_SIZE + 2) {
+    return Error
+  }
+  all_tx_data = { data, instruction_datas }
+  data_start_position = 2
+
+  for i in 0..num_signatures {
+      offsets = (Secp256r1SignatureOffsets) 
+        all_tx_data.data[data_start_position..data_start_position + SERIALIZED_OFFSET_STRUCT_SIZE]
+      data_position += SERIALIZED_OFFSET_STRUCT_SIZE
+
+      signature = get_data_slice(all_tx_data,
+                                offsets.signature_instruction_index,
+                                offsets.signature_offset
+                                signature_length)
+      if !signature {
+        return Error
+      }
+      if signature_S == highS {
+      return Error
+      }
+      public_key = get_data_slice(all_tx_data,
+                                  offsets.public_key_instruction_index,
+                                  offsets.public_key_offset,
+                                  SERIALIZED_PUBLIC_KEY_LENGTH)
+      if !public_key {
+        return Error
+      }
+
+      message = get_data_slice(all_tx_data,
+                              offsets.message_instruction_index,
+                              offsets.message_offset
+                              offsets.message_length)
+      if !message {
+        return Error
+      }
+
+      // sigverify includes validating signature and public_key
+      result = sigverify(signature, public_key, message)
+      if result != Success {
+        return Error
+      }
+    }
+    return Success
+}
+
+fn get_data_slice(all_tx_data, instruction_index, offset, length) {
```

Comment:
This should be reused across precompiles.

Reply:
added a comment to make re-use and reference to SIMD-0152 explicit 👍